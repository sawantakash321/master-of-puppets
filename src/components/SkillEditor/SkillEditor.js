import React, { Component } from 'react';
import { connect } from 'kea';
import classNames from 'classnames';
import { Select, Skills } from '../index';
import state from '../../utils/keaState';
import experiences from '../../utils/experienceItems';

@connect({
  actions: [state, ['setValue', 'submit', 'remove']],
  props: [state, ['values', 'isSubmitting', 'isLoading', 'skills']]
})
class SkillEditor extends Component {
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);
  }

  onSubmit(evt) {
    evt.preventDefault();
    const { actions } = this.props;
    const { submit } = actions;
    submit();
  }

  render() {
    const { values, isSubmitting, isLoading, skills, actions } = this.props;
    const validName =
      values.name && values.name.length >= 4 && values.name.length <= 255;
    const canSubmit =
      validName && values.experience && !isSubmitting && !isLoading;

    return (
      <div className="page">
        <form onSubmit={this.onSubmit}>
          <label htmlFor="skill" className="form-label">
            Add your skills
          </label>
          <div className="inline-form">
            <input
              id="skill"
              name="skill"
              type="text"
              placeholder="Node JS, Postgres, React, etc."
              value={values.name}
              onChange={e => actions.setValue('name', e.target.value)}
              minLength={4}
              maxLength={255}
              className={classNames({ invalid: !validName })}
            />
            <Select
              items={experiences}
              value={values.experience}
              id="experience"
              name="experience"
              placeholder="Experience"
              onChange={e => actions.setValue('experience', e.target.value)}
            />
            <button type="submit" className="btn" disabled={!canSubmit}>
              Add your skills
            </button>
          </div>
        </form>
        <Skills skills={skills} onRemove={actions.remove} />
      </div>
    );
  }
}

export default SkillEditor;
