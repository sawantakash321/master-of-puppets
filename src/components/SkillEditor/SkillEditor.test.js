/* global jest, describe, it, expect */
import React from 'react';
import ReactDOM from 'react-dom';
import { resetKeaCache, getStore } from 'kea';
import { compose } from 'redux';
import { Provider } from 'react-redux';
import sagaPlugin from 'kea-saga';
import renderer from 'react-test-renderer';
import logic from '../../utils/keaState';
import SkillEditor from './SkillEditor';

/* region Test setup*/
function getTestingStore() {
  resetKeaCache();
  return getStore({
    plugins: [sagaPlugin],
    paths: ['cw'],
    reducers: {},
    preloadedState: undefined,
    middleware: [],
    compose: compose,
    enhancers: []
  });
}

function getFakeStore() {
  return {
    dispatch: jest.fn(),
    subscribe: jest.fn(),
    getState: () => {
      return { cw: { values: {} } };
    }
  };
}

/* endregion */

describe('Design page', function() {
  it('renders without crashing', () => {
    const fakeStore = getFakeStore();
    const div = document.createElement('div');
    ReactDOM.render(<SkillEditor store={fakeStore} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('disable the submit button if there are no values on the form', () => {
    const fakeStore = getFakeStore();
    const design = mount(<SkillEditor store={fakeStore} />);
    const submit = design.find('button[type="submit"]');
    const submitNode = submit.getDOMNode();
    expect(submitNode.attributes.getNamedItem('disabled')).toBeTruthy();
  });

  // Snapshot test
  it('matches the snapshot', () => {
    const store = getFakeStore();

    const tree = renderer.create(<SkillEditor store={store} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
