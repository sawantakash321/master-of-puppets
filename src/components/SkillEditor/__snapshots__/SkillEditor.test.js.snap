// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Design page matches the snapshot 1`] = `
<div
  className="page"
>
  <form
    onSubmit={[Function]}
  >
    <label
      className="form-label"
      htmlFor="skill"
    >
      Add your skills
    </label>
    <div
      className="inline-form"
    >
      <input
        className="invalid"
        id="skill"
        maxLength={255}
        minLength={4}
        name="skill"
        onChange={[Function]}
        placeholder="Node JS, Postgres, React, etc."
        type="text"
        value={undefined}
      />
      <div
        className="select"
      >
        <select
          id="experience"
          name="experience"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          value={undefined}
        >
          <option
            className="placeholder"
            value=""
          >
            Experience
          </option>
          <option
            value="<1 year"
          >
            &lt;1 year
          </option>
          <option
            value="1-3 years"
          >
            1-3 years
          </option>
          <option
            value="3-5 years"
          >
            3-5 years
          </option>
          <option
            value="5-7 years"
          >
            5-7 years
          </option>
          <option
            value="7+ years"
          >
            7+ years
          </option>
        </select>
      </div>
      <button
        className="btn"
        disabled={true}
        type="submit"
      >
        Add your skills
      </button>
    </div>
  </form>
  <ul
    className="skills"
  />
</div>
`;
